claat/fetch/fetch.go

```diff
@@ -13,8 +13,419 @@
+//
+// The function will also fetch and parse fragments included
+// with types.ImportNode.
+func (f *Fetcher) SlurpCodelab(src string) (*codelab, error) {
+	res, err := f.fetch(src)
+	if err != nil {
+		return nil, err
+	}
+	defer res.body.Close()
+
+	opts := *parser.NewOptions()
+	opts.PassMetadata = f.passMetadata
+
+	clab, err := parser.Parse(string(res.typ), res.body, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	// fetch imports and parse them as fragments
+	var imports []*types.ImportNode
+	for _, st := range clab.Steps {
+		imports = append(imports, types.ImportNodes(st.Content.Nodes)...)
+	}
+	ch := make(chan error, len(imports))
+	defer close(ch)
+	for _, imp := range imports {
+		go func(n *types.ImportNode) {
+			frag, err := f.slurpFragment(n.URL)
+			if err != nil {
+				ch <- fmt.Errorf("%s: %v", n.URL, err)
+				return
+			}
+			n.Content.Nodes = frag
+			ch <- nil
+		}(imp)
+	}
+	for range imports {
+		if err := <-ch; err != nil {
+			return nil, err
+		}
+	}
+
+	v := &codelab{
+		Codelab: clab,
+		Typ:     res.typ,
+		Mod:     res.mod,
+	}
+	return v, nil
+}
+
+func (f *Fetcher) SlurpImages(src, dir string, steps []*types.Step) (map[string]string, error) {
+	// make sure img dir exists
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+
+	type res struct {
+		url, file string
+		err       error
+	}
+
+	ch := make(chan *res, 100)
+	defer close(ch)
+	var count int
+	for _, st := range steps {
+		nodes := types.ImageNodes(st.Content.Nodes)
+		count += len(nodes)
+		for _, n := range nodes {
+			go func(n *types.ImageNode) {
+				url := n.Src
+				file, err := f.slurpBytes(src, dir, url)
+				if err == nil {
+					n.Src = filepath.Join(util.ImgDirname, file)
+				}
+				ch <- &res{url, file, err}
+			}(n)
+		}
+	}
+
+	imap := make(map[string]string, count)
+	var err error
+	for i := 0; i < count; i++ {
+		r := <-ch
+		imap[r.file] = r.url
+		if r.err != nil && err == nil {
+			// record first error
+			err = fmt.Errorf("%s => %s: %v", r.url, r.file, r.err)
+		}
+	}
+
+	return imap, err
+}
+
+func (f *Fetcher) slurpBytes(codelabSrc, dir, imgURL string) (string, error) {
+	// images can be local in Markdown cases or remote.
+	// Only proceed a simple copy on local reference.
+	var b []byte
+	var ext string
+	u, err := url.Parse(imgURL)
+	if err != nil {
+		return "", err
+	}
+
+	// If the codelab source is being downloaded from the network, then we should interpret
+	// the image URL in the same way.
+	srcUrl, err := url.Parse(codelabSrc)
+	if err == nil && srcUrl.Host != "" {
+		u = srcUrl.ResolveReference(u)
+	}
+
+	if u.Host == "" {
+		if imgURL, err = restrictPathToParent(imgURL, filepath.Dir(codelabSrc)); err != nil {
+			return "", err
+		}
+		b, err = ioutil.ReadFile(imgURL)
+		ext = filepath.Ext(imgURL)
+	} else {
+		b, err = f.slurpRemoteBytes(u.String(), 5)
+		if string(b[6:10]) == "JFIF" {
+			ext = ".jpeg"
+		} else if string(b[0:3]) == "GIF" {
+			ext = ".gif"
+		} else {
+			ext = ".png"
```

Comment:
Should it also confirm that it's really a PNG, and error out otherwise?

Reply:
Out of scope for this PR I think. That's nice to have, but to the best of my recollection this hasn't been an issue in practice that I've heard of.